tests/returntypes/test_new_tape.py

```diff
@@ -0,0 +1,536 @@
+    def test_no_device_error(self, measurement):
+        """Test that an error is raised when a measurement that requires a device
+        is called without a device"""
+        msg = "The device argument is required to obtain the shape of the measurement process"
+
+        with pytest.raises(MeasurementShapeError, match=msg):
+            measurement.shape()
+
+    def test_undefined_shape_error(self):
+        """Test that an error is raised for a measurement with an undefined shape"""
+        measurement = qml.counts(wires=[0, 1])
+        msg = "Cannot deduce the shape of the measurement process with unrecognized return_type"
+
+        with pytest.raises(qml.QuantumFunctionError, match=msg):
+            measurement.shape()
+
+
+class TestOutputShape:
+    """Tests for determining the tape output shape of tapes."""
+
+    @pytest.mark.parametrize("measurement, expected_shape", measures)
+    @pytest.mark.parametrize("shots", [None, 1, 10])
+    def test_output_shapes_single(self, measurement, expected_shape, shots):
+        """Test that the output shape produced by the tape matches the expected
+        output shape."""
+        if shots is None and measurement.return_type is qml.measurements.Sample:
+            pytest.skip("Sample doesn't support analytic computations.")
+
+        num_wires = 3
+        dev = qml.device("default.qubit", wires=num_wires, shots=shots)
+
+        a = np.array(0.1)
+        b = np.array(0.2)
+
+        with qml.tape.QuantumTape() as tape:
+            qml.RY(a, wires=0)
+            qml.RX(b, wires=0)
+            qml.apply(measurement)
+
+        shot_dim = shots if not isinstance(shots, tuple) else len(shots)
+        if expected_shape is None:
+            expected_shape = shot_dim if shot_dim == 1 else (shot_dim,)
+
+        if measurement.return_type is qml.measurements.Sample:
+            if measurement.obs is not None:
+                expected_shape = () if shots == 1 else (shots,)
+            else:
+                expected_shape = (num_wires,) if shots == 1 else (shots, num_wires)
+
+        assert tape.shape(dev) == expected_shape
+
+    @pytest.mark.parametrize("measurement, expected_shape", measures)
+    @pytest.mark.parametrize("shots", [None, 1, 10, (1, 2, 5, 3)])
+    def test_output_shapes_single_qnode_check(self, measurement, expected_shape, shots):
+        """Test that the output shape produced by the tape matches the output
+        shape of a QNode for a single measurement."""
+        if shots is None and measurement.return_type is qml.measurements.Sample:
+            pytest.skip("Sample doesn't support analytic computations.")
+
+        warning = None
+        if shots is not None and measurement.return_type is qml.measurements.State:
+            # this is allowed by the tape but raises a warning
+            warning = (UserWarning, "Requested state or density matrix with finite shots")
```

Comment:
How about merging this into the logic starting on line 198 to group it into a single block?